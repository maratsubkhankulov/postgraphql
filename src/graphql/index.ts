import createGraphQLSchema from './schema/createGQLSchema'
import createGraphQLHTTPRequestHandler from './http/createGraphQLHTTPRequestHandler'

export {
  createGraphQLSchema,
  createGraphQLHTTPRequestHandler,
}
