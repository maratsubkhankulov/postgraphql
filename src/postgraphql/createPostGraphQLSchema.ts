import { Client, connect } from 'pg'
import { GraphQLSchema } from 'graphql'
import { Inventory } from '../interface'
import { introspectPGDatabase, addPGCatalogToInventory } from '../postgres'
import { createGraphQLSchema } from '../graphql'
import BuildToken from '../graphql/schema/BuildToken'
import createPGProcedureMutationGQLFieldEntries from './procedures/hooks/createPGProcedureMutationGQLFieldEntries'
import createPGProcedureQueryGQLFieldEntries from './procedures/hooks/createPGProcedureQueryGQLFieldEntries'

/**
 * Creates a PostGraphQL schema by looking at a Postgres client.
 */
export default async function createPostGraphQLSchema (
  config: {
    // TODO: Accept connection strings, configs, clients, and `PGCatalog`s.
    pgClient: Client,
    pgSchemas?: Array<string>,
    relay1Ids?: boolean,
  },
): Promise<GraphQLSchema> {
  // Create our inventory.
  const inventory = new Inventory()

  // Introspect our Postgres database to get a catalog.
  const pgCatalog = await introspectPGDatabase(
    config.pgClient,
    config.pgSchemas || ['public'],
  )

  // Add all of our Postgres constructs to that inventory.
  addPGCatalogToInventory(inventory, pgCatalog, {
    renameIdToRowId: config.relay1Ids,
  })

  // Actually create our GraphQL schema.
  const gqlSchema = createGraphQLSchema(inventory, {
    nodeIdFieldName: config.relay1Ids ? 'id' : '__id',
    _hooks: {
      mutationFieldEntries: (_buildToken: BuildToken) => createPGProcedureMutationGQLFieldEntries(_buildToken, pgCatalog),
      queryFieldEntries: (_buildToken: BuildToken) => createPGProcedureQueryGQLFieldEntries(_buildToken, pgCatalog),
      objectTypeFieldEntries: () => [],
    },
  })

  return gqlSchema
}
