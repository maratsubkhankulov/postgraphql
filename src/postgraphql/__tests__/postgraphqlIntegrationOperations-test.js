import { resolve as resolvePath } from 'path'
import { readFile, readdirSync } from 'fs'
import { graphql } from 'graphql'
import withPGClient from '../../postgres/__tests__/fixtures/withPGClient'
import { $$pgClient } from '../../postgres/inventory/pgClientFromContext'
import createPostGraphQLSchema from '../createPostGraphQLSchema'

const kitchenSinkData = new Promise((resolve, reject) => {
  readFile(resolvePath(__dirname, '../../../resources/kitchen-sink-data.sql'), (error, data) => {
    if (error) reject(error)
    else resolve(data.toString())
  })
})

const queriesDir = resolvePath(__dirname, 'fixtures/queries')

for (const file of readdirSync(queriesDir)) {
  test(`operation ${file}`, withPGClient(async pgClient => {
    const gqlSchema = await createPostGraphQLSchema({
      pgClient,
      pgSchemas: ['a', 'b', 'c'],
      relay1Ids: false,
    })

    const query = await new Promise((resolve, reject) => {
      readFile(resolvePath(queriesDir, file), (error, data) => {
        if (error) reject(error)
        else resolve(data.toString())
      })
    })

    await pgClient.query(await kitchenSinkData)

    const result = await graphql(gqlSchema, query, null, { [$$pgClient]: pgClient })

    // Log the errors in our result.
    if (result.errors)
      for (const e of result.errors)
        console.error(e.stack || e)

    expect(result).toMatchSnapshot()
  }))
}
