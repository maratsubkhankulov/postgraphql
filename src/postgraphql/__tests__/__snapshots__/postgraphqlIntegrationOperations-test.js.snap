exports[`test operation connections.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxXV0=",
          "node": Object {
            "email": "john.smith@email.com",
            "id": 1,
            "name": "John Smith",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsyXV0=",
          "node": Object {
            "email": "sara.smith@email.com",
            "id": 2,
            "name": "Sara Smith",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFszXV0=",
          "node": Object {
            "email": "budd.deey@email.com",
            "id": 3,
            "name": "Budd Deey",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs0XV0=",
          "node": Object {
            "email": "kathryn.ramirez@email.com",
            "id": 4,
            "name": "Kathryn Ramirez",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs1XV0=",
          "node": Object {
            "email": "johnny.tucker@email.com",
            "id": 5,
            "name": "Johnny Tucker",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 5,
    },
    "b": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxXV0=",
          "node": Object {
            "email": "john.smith@email.com",
            "id": 1,
            "name": "John Smith",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsyXV0=",
          "node": Object {
            "email": "sara.smith@email.com",
            "id": 2,
            "name": "Sara Smith",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": true,
        "hasPreviousPage": false,
      },
      "totalCount": 5,
    },
    "c": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs0XV0=",
          "node": Object {
            "email": "kathryn.ramirez@email.com",
            "id": 4,
            "name": "Kathryn Ramirez",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs1XV0=",
          "node": Object {
            "email": "johnny.tucker@email.com",
            "id": 5,
            "name": "Johnny Tucker",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": true,
      },
      "totalCount": 5,
    },
    "d": Object {
      "edges": Array [
        Object {
          "cursor": "WyJuYW1lX2FzYyIsWyJCdWRkIERlZXkiLDNdXQ==",
          "node": Object {
            "email": "budd.deey@email.com",
            "id": 3,
            "name": "Budd Deey",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2FzYyIsWyJKb2huIFNtaXRoIiwxXV0=",
          "node": Object {
            "email": "john.smith@email.com",
            "id": 1,
            "name": "John Smith",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2FzYyIsWyJKb2hubnkgVHVja2VyIiw1XV0=",
          "node": Object {
            "email": "johnny.tucker@email.com",
            "id": 5,
            "name": "Johnny Tucker",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2FzYyIsWyJLYXRocnluIFJhbWlyZXoiLDRdXQ==",
          "node": Object {
            "email": "kathryn.ramirez@email.com",
            "id": 4,
            "name": "Kathryn Ramirez",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2FzYyIsWyJTYXJhIFNtaXRoIiwyXV0=",
          "node": Object {
            "email": "sara.smith@email.com",
            "id": 2,
            "name": "Sara Smith",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 5,
    },
    "e": Object {
      "edges": Array [
        Object {
          "cursor": "WyJuYW1lX2Rlc2MiLFsiU2FyYSBTbWl0aCIsMl1d",
          "node": Object {
            "email": "sara.smith@email.com",
            "id": 2,
            "name": "Sara Smith",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2Rlc2MiLFsiS2F0aHJ5biBSYW1pcmV6Iiw0XV0=",
          "node": Object {
            "email": "kathryn.ramirez@email.com",
            "id": 4,
            "name": "Kathryn Ramirez",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2Rlc2MiLFsiSm9obm55IFR1Y2tlciIsNV1d",
          "node": Object {
            "email": "johnny.tucker@email.com",
            "id": 5,
            "name": "Johnny Tucker",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2Rlc2MiLFsiSm9obiBTbWl0aCIsMV1d",
          "node": Object {
            "email": "john.smith@email.com",
            "id": 1,
            "name": "John Smith",
          },
        },
        Object {
          "cursor": "WyJuYW1lX2Rlc2MiLFsiQnVkZCBEZWV5IiwzXV0=",
          "node": Object {
            "email": "budd.deey@email.com",
            "id": 3,
            "name": "Budd Deey",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 5,
    },
    "f": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxXV0=",
          "node": Object {
            "email": "john.smith@email.com",
            "id": 1,
            "name": "John Smith",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": true,
        "hasPreviousPage": false,
      },
      "totalCount": 5,
    },
    "g": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFszXV0=",
          "node": Object {
            "email": "budd.deey@email.com",
            "id": 3,
            "name": "Budd Deey",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs0XV0=",
          "node": Object {
            "email": "kathryn.ramirez@email.com",
            "id": 4,
            "name": "Kathryn Ramirez",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs1XV0=",
          "node": Object {
            "email": "johnny.tucker@email.com",
            "id": 5,
            "name": "Johnny Tucker",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": true,
      },
      "totalCount": 5,
    },
    "h": Object {
      "edges": Array [
        Object {
          "cursor": "WyJuYXR1cmFsIiwxXQ==",
          "node": Object {
            "constant": 2,
            "name": "John Smith",
            "x": 1,
          },
        },
        Object {
          "cursor": "WyJuYXR1cmFsIiwyXQ==",
          "node": Object {
            "constant": 2,
            "name": "Sara Smith",
            "x": 2,
          },
        },
        Object {
          "cursor": "WyJuYXR1cmFsIiwzXQ==",
          "node": Object {
            "constant": 2,
            "name": "Budd Deey",
            "x": 3,
          },
        },
        Object {
          "cursor": "WyJuYXR1cmFsIiw0XQ==",
          "node": Object {
            "constant": 2,
            "name": "Kathryn Ramirez",
            "x": 4,
          },
        },
        Object {
          "cursor": "WyJuYXR1cmFsIiw1XQ==",
          "node": Object {
            "constant": 2,
            "name": "Johnny Tucker",
            "x": 5,
          },
        },
      ],
    },
    "i": Object {
      "edges": Array [
        Object {
          "cursor": "WyJjb25zdGFudF9hc2MiLFsyXV0=",
          "node": Object {
            "constant": 2,
            "name": "John Smith",
            "x": 1,
          },
        },
        Object {
          "cursor": "WyJjb25zdGFudF9hc2MiLFsyXV0=",
          "node": Object {
            "constant": 2,
            "name": "Sara Smith",
            "x": 2,
          },
        },
        Object {
          "cursor": "WyJjb25zdGFudF9hc2MiLFsyXV0=",
          "node": Object {
            "constant": 2,
            "name": "Budd Deey",
            "x": 3,
          },
        },
        Object {
          "cursor": "WyJjb25zdGFudF9hc2MiLFsyXV0=",
          "node": Object {
            "constant": 2,
            "name": "Kathryn Ramirez",
            "x": 4,
          },
        },
        Object {
          "cursor": "WyJjb25zdGFudF9hc2MiLFsyXV0=",
          "node": Object {
            "constant": 2,
            "name": "Johnny Tucker",
            "x": 5,
          },
        },
      ],
    },
    "j": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxXV0=",
          "node": Object {
            "authorId": 2,
            "headline": "No… It’s a thing; it’s like a plan, but with more greatness.",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs4XV0=",
          "node": Object {
            "authorId": 2,
            "headline": "It’s a fez. I wear a fez now. Fezes are cool.",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxMF1d",
          "node": Object {
            "authorId": 2,
            "headline": "What’s with you kids? Every other day it’s food, food, food.",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": false,
      },
      "totalCount": 3,
    },
    "k": Object {
      "edges": Array [
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFsxXV0=",
          "node": Object {
            "authorId": 2,
            "headline": "No… It’s a thing; it’s like a plan, but with more greatness.",
          },
        },
        Object {
          "cursor": "WyJwcmltYXJ5X2tleV9hc2MiLFs4XV0=",
          "node": Object {
            "authorId": 2,
            "headline": "It’s a fez. I wear a fez now. Fezes are cool.",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": true,
        "hasPreviousPage": false,
      },
      "totalCount": 3,
    },
    "l": Object {
      "edges": Array [
        Object {
          "cursor": "WyJoZWFkbGluZV9hc2MiLFsiWW91IGhpdCBtZSB3aXRoIGEgY3JpY2tldCBiYXQuIiw0XV0=",
          "node": Object {
            "authorId": 1,
            "headline": "You hit me with a cricket bat.",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
        "hasPreviousPage": true,
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`test operation mutation-create.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "type": Object {
        "__id": "WyJ0eXBlcyIsMjAxXQ==",
        "bigint": 50,
        "boolean": false,
        "compoundType": Object {
          "a": 123,
          "b": "abc",
          "c": "GREEN",
          "d": "ec4a9fae-4ec5-4763-98eb-0327fb2dc9bf",
          "fooBar": 20,
        },
        "domain": 6,
        "domain2": 5,
        "enum": "RED",
        "id": 201,
        "nestedCompoundType": Object {
          "a": Object {
            "a": 456,
            "b": "def",
            "c": "BLUE",
            "d": "79863dcf-0433-4c3d-bc51-978326d4546f",
            "fooBar": 42,
          },
          "b": Object {
            "a": 789,
            "b": "ghi",
            "c": "RED",
            "d": "b687ee42-c515-4544-b742-525e39517e7d",
            "fooBar": -8,
          },
          "bazBuz": 0,
        },
        "varchar": "abc",
      },
    },
    "b": Object {
      "a": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "b": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "c": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "clientMutationId": null,
      "d": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "e": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "f": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "g": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
      },
      "person": Object {
        "__id": "WyJwZW9wbGUiLDkwMDBd",
        "about": "Son of Sara and John Smith.",
        "email": "johnny.boy.smith@email.com",
        "id": 9000,
        "name": "John Smith Jr.",
      },
    },
    "c": Object {
      "a": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "b": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "c": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "clientMutationId": "hello",
      "d": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "e": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "f": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "g": Object {
        "cursor": null,
        "node": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
      },
      "person": Object {
        "__id": "WyJwZW9wbGUiLDIwXQ==",
        "about": "My archnemisis is Budd Deey.",
        "email": "best.pal@email.com",
        "id": 20,
        "name": "Best Pal",
      },
    },
    "d": Object {
      "clientMutationId": "world",
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiw5MDAwLDIwXQ==",
        "extra": false,
        "personByPersonId1": Object {
          "__id": "WyJwZW9wbGUiLDkwMDBd",
          "name": "John Smith Jr.",
        },
        "personByPersonId2": Object {
          "__id": "WyJwZW9wbGUiLDIwXQ==",
          "name": "Best Pal",
        },
        "personId1": 9000,
        "personId2": 20,
      },
    },
  },
}
`;

exports[`test operation mutation-delete.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "deletedPostId": "WyJwb3N0cyIsMV0=",
      "post": Object {
        "__id": "WyJwb3N0cyIsMV0=",
        "authorId": 2,
        "headline": "No… It’s a thing; it’s like a plan, but with more greatness.",
        "id": 1,
      },
    },
    "b": Object {
      "clientMutationId": "hello",
      "deletedPostId": "WyJwb3N0cyIsMl0=",
      "post": Object {
        "__id": "WyJwb3N0cyIsMl0=",
        "authorId": 1,
        "headline": "I hate yogurt. It’s just stuff with bits in.",
        "id": 2,
      },
    },
    "c": null,
    "d": Object {
      "clientMutationId": "world",
      "deletedPostId": "WyJwb3N0cyIsM10=",
      "post": Object {
        "__id": "WyJwb3N0cyIsM10=",
        "authorId": 1,
        "headline": "Is that a cooking show?",
        "id": 3,
      },
    },
    "e": Object {
      "clientMutationId": null,
      "deletedPostId": "WyJwb3N0cyIsNl0=",
      "post": Object {
        "__id": "WyJwb3N0cyIsNl0=",
        "authorId": 3,
        "headline": "Stop talking, brain thinking. Hush.",
        "id": 6,
      },
    },
    "f": Object {
      "clientMutationId": "hello",
      "deletedPostId": "WyJwb3N0cyIsOV0=",
      "post": Object {
        "__id": "WyJwb3N0cyIsOV0=",
        "authorId": 3,
        "headline": "You know how I sometimes have really brilliant ideas?",
        "id": 9,
      },
    },
    "g": null,
    "h": Object {
      "clientMutationId": "world",
      "deletedPostId": "WyJwb3N0cyIsMTFd",
      "post": Object {
        "__id": "WyJwb3N0cyIsMTFd",
        "authorId": 3,
        "headline": "They’re not aliens, they’re Earth…liens!",
        "id": 11,
      },
    },
    "i": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
        "personByPersonId1": Object {
          "__id": "WyJwZW9wbGUiLDRd",
          "name": "Kathryn Ramirez",
        },
        "personByPersonId2": Object {
          "__id": "WyJwZW9wbGUiLDNd",
          "name": "Budd Deey",
        },
        "personId1": 4,
        "personId2": 3,
      },
      "deletedCompoundKeyId": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
    },
    "j": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
        "personByPersonId1": Object {
          "__id": "WyJwZW9wbGUiLDJd",
          "name": "Sara Smith",
        },
        "personByPersonId2": Object {
          "__id": "WyJwZW9wbGUiLDNd",
          "name": "Budd Deey",
        },
        "personId1": 2,
        "personId2": 3,
      },
      "deletedCompoundKeyId": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
    },
    "k": Object {
      "clientMutationId": null,
      "deletedPersonId": "WyJwZW9wbGUiLDNd",
      "person": Object {
        "__id": "WyJwZW9wbGUiLDNd",
        "email": "budd.deey@email.com",
        "id": 3,
        "name": "Budd Deey",
      },
    },
  },
  "errors": Array [
    [GraphQLError: No values were deleted in collection 'posts' because no values were found.],
    [GraphQLError: No values were deleted in collection 'posts' because no values were found.],
  ],
}
`;

exports[`test operation mutation-update.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "clientMutationId": null,
      "person": Object {
        "__id": "WyJwZW9wbGUiLDFd",
        "about": "An older John Smith",
        "email": "john.smith@email.com",
        "id": 1,
        "name": "John Smith Sr.",
      },
    },
    "b": Object {
      "clientMutationId": "hello",
      "person": Object {
        "__id": "WyJwZW9wbGUiLDJd",
        "about": null,
        "email": "sarah.smith@email.com",
        "id": 2,
        "name": "Sarah Smith",
      },
    },
    "c": Object {
      "clientMutationId": "world",
      "person": Object {
        "__id": "WyJwZW9wbGUiLDJd",
        "about": "Now with an “H.”",
        "email": "sarah.smith@email.com",
        "id": 2,
        "name": "Sarah Smith",
      },
    },
    "d": Object {
      "clientMutationId": null,
      "person": Object {
        "__id": "WyJwZW9wbGUiLDNd",
        "about": "I have taken over Budd’s account. Hehehe.",
        "email": "budd.deey@email.com",
        "id": 3,
        "name": "Best Pal",
      },
    },
    "e": Object {
      "clientMutationId": null,
      "person": Object {
        "__id": "WyJwZW9wbGUiLDRd",
        "about": "Make art not friends.",
        "email": "kathryn.ramirez@email.com",
        "id": 4,
        "name": "Kathryn Ramirez",
      },
    },
    "f": Object {
      "clientMutationId": null,
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDJd",
        "extra": true,
        "personByPersonId1": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personByPersonId2": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personId1": 2,
        "personId2": 2,
      },
    },
    "g": Object {
      "clientMutationId": "hello",
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiwzLDJd",
        "extra": false,
        "personByPersonId1": Object {
          "id": 3,
          "name": "Best Pal",
        },
        "personByPersonId2": Object {
          "id": 2,
          "name": "Sarah Smith",
        },
        "personId1": 3,
        "personId2": 2,
      },
    },
    "h": Object {
      "clientMutationId": "world",
      "compoundKey": Object {
        "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
        "extra": false,
        "personByPersonId1": Object {
          "id": 4,
          "name": "Kathryn Ramirez",
        },
        "personByPersonId2": Object {
          "id": 3,
          "name": "Best Pal",
        },
        "personId1": 4,
        "personId2": 3,
      },
    },
  },
}
`;

exports[`test operation node.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "nodes": Array [
        Object {
          "__id": "WyJwZW9wbGUiLDFd",
          "__typename": "Person",
          "id": 1,
          "name": "John Smith",
        },
        Object {
          "__id": "WyJwZW9wbGUiLDJd",
          "__typename": "Person",
          "id": 2,
          "name": "Sara Smith",
        },
        Object {
          "__id": "WyJwZW9wbGUiLDNd",
          "__typename": "Person",
          "id": 3,
          "name": "Budd Deey",
        },
        Object {
          "__id": "WyJwZW9wbGUiLDRd",
          "__typename": "Person",
          "id": 4,
          "name": "Kathryn Ramirez",
        },
        Object {
          "__id": "WyJwZW9wbGUiLDVd",
          "__typename": "Person",
          "id": 5,
          "name": "Johnny Tucker",
        },
      ],
    },
    "b": Object {
      "nodes": Array [
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
          "__typename": "CompoundKey",
          "personId1": 1,
          "personId2": 2,
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDFd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 1,
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 3,
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDVd",
          "__typename": "CompoundKey",
          "personId1": 2,
          "personId2": 5,
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDNd",
          "__typename": "CompoundKey",
          "personId1": 4,
          "personId2": 3,
        },
        Object {
          "__id": "WyJjb21wb3VuZF9rZXlzIiw0LDRd",
          "__typename": "CompoundKey",
          "personId1": 4,
          "personId2": 4,
        },
      ],
    },
    "c": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
      "__typename": "CompoundKey",
      "personId1": 2,
      "personId2": 3,
    },
    "d": Object {
      "__id": "WyJwZW9wbGUiLDJd",
      "__typename": "Person",
      "id": 2,
      "name": "Sara Smith",
    },
    "e": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
      "__typename": "CompoundKey",
      "personId1": 1,
      "personId2": 2,
    },
    "f": Object {
      "__id": "WyJwZW9wbGUiLDVd",
      "__typename": "Person",
      "id": 5,
      "name": "Johnny Tucker",
    },
    "g": null,
    "h": null,
    "i": Object {
      "__id": "WyJwZW9wbGUiLDJd",
      "__typename": "Person",
      "id": 2,
      "name": "Sara Smith",
    },
    "j": Object {
      "__id": "WyJwZW9wbGUiLDVd",
      "__typename": "Person",
      "id": 5,
      "name": "Johnny Tucker",
    },
    "k": null,
    "l": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwyLDNd",
      "__typename": "CompoundKey",
      "personId1": 2,
      "personId2": 3,
    },
    "m": Object {
      "__id": "WyJjb21wb3VuZF9rZXlzIiwxLDJd",
      "__typename": "CompoundKey",
      "personId1": 1,
      "personId2": 2,
    },
    "n": null,
  },
}
`;

exports[`test operation query.graphql 1`] = `
Object {
  "data": Object {
    "node": null,
    "query": Object {
      "node": null,
      "query": Object {
        "node": null,
        "query": Object {
          "node": null,
        },
      },
    },
  },
}
`;

exports[`test operation relation-tail-head.graphql 1`] = `
Object {
  "data": Object {
    "allCompoundKeys": Object {
      "nodes": Array [
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "email": "john.smith@email.com",
            "name": "John Smith",
          },
          "personByPersonId2": Object {
            "email": "sara.smith@email.com",
            "name": "Sara Smith",
          },
          "personId1": 1,
          "personId2": 2,
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "email": "sara.smith@email.com",
            "name": "Sara Smith",
          },
          "personByPersonId2": Object {
            "email": "john.smith@email.com",
            "name": "John Smith",
          },
          "personId1": 2,
          "personId2": 1,
        },
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "email": "sara.smith@email.com",
            "name": "Sara Smith",
          },
          "personByPersonId2": Object {
            "email": "budd.deey@email.com",
            "name": "Budd Deey",
          },
          "personId1": 2,
          "personId2": 3,
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "email": "sara.smith@email.com",
            "name": "Sara Smith",
          },
          "personByPersonId2": Object {
            "email": "johnny.tucker@email.com",
            "name": "Johnny Tucker",
          },
          "personId1": 2,
          "personId2": 5,
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez",
          },
          "personByPersonId2": Object {
            "email": "budd.deey@email.com",
            "name": "Budd Deey",
          },
          "personId1": 4,
          "personId2": 3,
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez",
          },
          "personByPersonId2": Object {
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez",
          },
          "personId1": 4,
          "personId2": 4,
        },
      ],
    },
    "allForeignKeys": Object {
      "nodes": Array [
        Object {
          "compoundKey1": 2,
          "compoundKey2": 1,
          "compoundKeyByCompoundKey1AndCompoundKey2": Object {
            "extra": false,
            "personId1": 2,
            "personId2": 1,
          },
          "personByPersonId": Object {
            "email": "johnny.tucker@email.com",
            "name": "Johnny Tucker",
          },
          "personId": 5,
        },
        Object {
          "compoundKey1": 4,
          "compoundKey2": 4,
          "compoundKeyByCompoundKey1AndCompoundKey2": Object {
            "extra": false,
            "personId1": 4,
            "personId2": 4,
          },
          "personByPersonId": null,
          "personId": null,
        },
        Object {
          "compoundKey1": 2,
          "compoundKey2": 1,
          "compoundKeyByCompoundKey1AndCompoundKey2": Object {
            "extra": false,
            "personId1": 2,
            "personId2": 1,
          },
          "personByPersonId": null,
          "personId": null,
        },
      ],
    },
  },
}
`;

exports[`test operation types.graphql 1`] = `
Object {
  "data": Object {
    "allTypes": Object {
      "nodes": Array [
        Object {
          "bigint": 200,
          "boolean": true,
          "compoundType": Object {
            "a": 1,
            "b": "2",
            "c": "BLUE",
            "d": "4be8a712-3ff7-432e-aa34-fdb43fbd838d",
            "fooBar": 8,
          },
          "domain": 5,
          "domain2": 6,
          "enum": "GREEN",
          "nestedCompoundType": Object {
            "a": Object {
              "a": 3,
              "b": "456",
              "c": "RED",
              "d": "aed18400-2a92-46df-8204-b70c728b3520",
              "fooBar": 93,
            },
            "b": Object {
              "a": 42,
              "b": "Hello, world!",
              "c": "BLUE",
              "d": "d34df5e0-83f1-11e6-8dd0-abee917ffd1e",
              "fooBar": -5,
            },
            "bazBuz": 7,
          },
        },
      ],
    },
  },
}
`;

exports[`test operation unique-constraints.graphql 1`] = `
Object {
  "data": Object {
    "a": Object {
      "email": "sara.smith@email.com",
      "name": "Sara Smith",
    },
    "b": null,
    "c": Object {
      "personId1": 1,
      "personId2": 2,
    },
    "d": Object {
      "personId1": 4,
      "personId2": 4,
    },
    "e": null,
  },
}
`;
